compendium/DeclarativeServices/src/manager/ConfigurationNotifier.cpp

```diff
@@ -0,0 +1,216 @@
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  =============================================================================*/
+
+#include "ConfigurationNotifier.hpp"
+#include "../ComponentRegistry.hpp"
+#include "../metadata/ComponentMetadata.hpp"
+#include "ComponentConfigurationImpl.hpp"
+#include "ComponentManagerImpl.hpp"
+#include "cppmicroservices/SharedLibraryException.h"
+#include "cppmicroservices/cm/ConfigurationAdmin.hpp"
+
+namespace cppmicroservices {
+namespace scrimpl {
+
+using cppmicroservices::scrimpl::metadata::ComponentMetadata;
+
+ConfigurationNotifier::ConfigurationNotifier(
+  cppmicroservices::BundleContext context,
+  std::shared_ptr<cppmicroservices::logservice::LogService> logger)
+  : tokenCounter(0)
+  , bundleContext(std::move(context))
+  , logger(std::move(logger))
+{
+  if (!bundleContext || !(this->logger)) {
+    throw std::invalid_argument("ConfigurationNotifier Constructor "
+                                "provided with invalid arguments");
+  }
+}
+
+cppmicroservices::ListenerTokenId ConfigurationNotifier::RegisterListener(
+  const std::string& pid,
+  std::function<void(const ConfigChangeNotification&)> notify,
+  std::shared_ptr<ComponentConfigurationImpl> mgr)
+{
+  cppmicroservices::ListenerTokenId retToken = ++tokenCounter;
+
+  {
+    auto listener = Listener(notify, mgr);
+
+    auto listenersMapHandle = listenersMap.lock();
+    auto iter = listenersMapHandle->find(pid);
+    if (iter != listenersMapHandle->end()) {
+      (*(iter->second)).emplace(retToken, listener);
+    } else {
+      auto tokenMapPtr = std::make_shared<TokenMap>();
+      tokenMapPtr->emplace(retToken, listener);
+      listenersMapHandle->emplace(pid, tokenMapPtr);
+    }
+  }
+
+  return retToken;
+}
+
+void ConfigurationNotifier::UnregisterListener(
+  const std::string& pid,
+  const cppmicroservices::ListenerTokenId token) noexcept
+{
+  auto listenersMapHandle = listenersMap.lock();
+  if (listenersMapHandle->empty() || pid.empty()) {
+    return;
+  }
+
+  auto iter = listenersMapHandle->find(pid);
+  if (iter == listenersMapHandle->end()) {
+    return;
+  }
+
+  auto tokenMapPtr = iter->second;
+  for (const auto& tokenMap : (*tokenMapPtr)) {
+    if (tokenMap.first == token) {
+      tokenMapPtr->erase(tokenMap.first);
+      if (tokenMapPtr->size() == 0) {
+        listenersMapHandle->erase(iter);
+      }
+      break;
+    }
+  }
+}
+bool ConfigurationNotifier::AnyListenersForPid(const std::string& pid) noexcept
+{
+  std::string factoryName;
+  auto mgr = std::shared_ptr<ComponentConfigurationImpl>();
```

Comment:
Jeff: Can initialize mgr this way: `std::shared_ptr<ComponentConfigurationImpl> mgr;`